src/group_theory/monoid_localization.lean

```diff
@@ -266,25 +274,295 @@ show _ * _ = _, by rw [mul_inv_left, mul_one]
+noncomputable def map : N →* Q :=
+@lift _ _ _ _ _ _ _ f (k.1.comp g) $ λ y, k.2 ⟨g y, hy y⟩
+
+variables {k}
+
+@[to_additive] lemma map_of (x) :
+  f.map hy k (f.1 x) = k.1 (g x) := f.lift_of (λ y, k.2 ⟨g y, hy y⟩) x
+
+@[simp, to_additive] lemma map_comp_of :
+  (f.map hy k).comp f.1 = k.1.comp g := f.lift_comp_of $ λ y, k.2 ⟨g y, hy y⟩
+
+@[to_additive] lemma map_mk (x) (y : S) :
+  f.map hy k (f.mk' x y) = k.mk' (g x) ⟨g y, hy y⟩ :=
+begin
+  rw [map, lift_mk', mul_inv_left],
+  { show k.1 (g x) = k.1 (g y) * _,
+    rw mul_mk'_eq_mk'_of_mul,
+    exact (k.mk'_mul_cancel_left (g x) ⟨(g y), hy y⟩).symm },
+end
+
+@[to_additive] lemma map_spec (z u) :
+  f.map hy k z = u ↔ k.1 (g (sec S f.1 z).1) = k.1 (g (sec S f.1 z).2) * u :=
+f.lift_spec (λ y, k.2 ⟨g y, hy y⟩) _ _
+
+@[to_additive] lemma map_mul_right (z) :
+  f.map hy k z * (k.1 (g (sec S f.1 z).2)) = k.1 (g (sec S f.1 z).1) :=
+f.lift_mul_right (λ y, k.2 ⟨g y, hy y⟩) _
+
+@[to_additive] lemma map_mul_left (z) :
+  k.1 (g (sec S f.1 z).2) * f.map hy k z = k.1 (g (sec S f.1 z).1) :=
+by rw [mul_comm, f.map_mul_right]
+
+@[simp, to_additive] lemma map_id (z : N) :
+  f.map (λ y, show monoid_hom.id M y ∈ S, from y.2) f z = z :=
+f.lift_id z
+
+@[to_additive] lemma map_comp_map {U : submonoid Q} {R} [comm_monoid R]
+  (j : localization_map U R) {l : P →* Q} (hl : ∀ w : T, l w ∈ U) :
+  (k.map hl j).comp (f.map hy k) = f.map (λ x, show l.comp g x ∈ U, from hl ⟨g x, hy x⟩) j :=
+begin
+  ext z,
+  show j.1 _ * _ = j.1 (l _) * _,
+  { rw [mul_inv_left, ←mul_assoc, mul_inv_right],
+    show j.1 _ * j.1 (l (g _)) = j.1 (l _) * _,
+    rw [←j.1.map_mul, ←j.1.map_mul, ←l.map_mul, ←l.map_mul],
+    exact comp_eq_of_eq k hl j
+      (by rw [k.1.map_mul, k.1.map_mul, sec_spec' k.3, mul_assoc, map_mul_right]) },
+end
+
+@[to_additive] lemma map_map {U : submonoid Q} {R} [comm_monoid R]
+  (j : localization_map U R) {l : P →* Q} (hl : ∀ w : T, l w ∈ U) (x) :
+  k.map hl j (f.map hy k x) = f.map (λ x, show l.comp g x ∈ U, from hl ⟨g x, hy x⟩) j x :=
+by rw ←f.map_comp_map hy j hl; refl
+
+variables {g}
+
+/-- If `f : M →* N` and `k : M →* R` are localization maps for a submonoid `S`, we get an
+    isomorphism of `N` and `R`. -/
+@[to_additive "If `f : M →+ N` and `k : M →+ R` are localization maps for a submonoid `S`, we get an isomorphism of `N` and `R`."]
+noncomputable def mul_equiv_of_localizations
+  (k : localization_map S P) : N ≃* P :=
+⟨f.lift k.2, k.lift f.2, f.lift_inv, k.lift_inv, monoid_hom.map_mul _⟩
+
+/-- If `f : M →* N` is a localization map for a submonoid `S` and `k : N ≃* P` is an isomorphism
+    of `comm_monoid`s, `k ∘ f` is a localization map for `M` at `S`. -/
+@[to_additive "If `f : M →+ N` is a localization map for a submonoid `S` and `k : N ≃+ P` is an isomorphism of `add_comm_monoid`s, `k ∘ f` is a localization map for `M` at `S`."]
+def mul_equiv_map (k : N ≃* P) :
```

Comment:
I think the following definitions should come with some compatibility lemmas. E.g., the `coe` to function of this `mul_equiv_map` is `k ∘ f.1`. But also, what happens if I compose this with `k.symm`: then I recover `f`. Similar lemmas apply below.

Reply:
@101damnations What do you think of :arrow_up: remark?

Reply:
added some in [this](https://github.com/leanprover-community/mathlib/pull/2118/commits/9dc2f5b372f5f6e361720eb3b48c870619d65ed5) commit and the latest commit but I might be missing some obvious ones, I'm not sure